hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/SCMBlockDeletingService.java

```diff
@@ -273,4 +279,24 @@ public void stop() {
   public ScmBlockDeletingServiceMetrics getMetrics() {
     return this.metrics;
   }
+
+  /**
+   * Filters and returns a set of healthy datanodes that have not exceeded
+   * the deleteBlocksPendingCommandLimit.
+   *
+   * @param datanodes a list of DatanodeDetails
+   * @return a set of filtered DatanodeDetails
+   */
+  @VisibleForTesting
+  protected Set<DatanodeDetails> getDatanodesWithinCommandLimit(
+      List<DatanodeDetails> datanodes) throws NodeNotFoundException {
+    final Set<DatanodeDetails> included = new HashSet<>();
+    for (DatanodeDetails dn : datanodes) {
+      if (nodeManager.getTotalDatanodeCommandCount(dn,
```

Comment:
The basic idea looks good, but I have a concern that the queued cmd size from the last dn heartbeat has a potential delay.  If the delete operation is frequent, there might be a case that the service cannot find idle data nodes to send cmd until the next heartbeat comes. The deletion process might not be smooth as you expect.
Pls correct me if i am wrong.

Reply:
I see. Most of the time it should be fine.
The worst case: 
A                      If DN reported the HB at time A with a full cmd queue. 
A+ 29.9s.        SCM DeletedBlockTransactionScanner executes and cannot send cmd to DN, needs to wait for next 
                        round.
A+ 90s.           SCM has updated the latest HB from DN and DeletedBlockTransactionScanner executes, and finally 
                        send cmd to DN again.
It could lead to at most a 90-sec gap. 
Right now it seems to be trivial compared to the interval of DeletedBlockTransactionScanner.  
Thx for the explanation.

Reply:
Yes, and when A+ 29.9s ~ A+ 90s the DN can continue to process the `DeleteBlocksCommand` in its command queue. And the DN's command queue is full, DN will not be idle. Because this PR is determine whether to continue sending `DeleteBlocksCommand` to DN according to the command queue length of DN.

Reply:
@xichen01 
1. Without processing DN HB, this will get same command again. So it is not useful to keep queue at SCM side. Else SCM will add duplicate command only, and DN will be processing duplicate commands only.
2. DN queue is limited to "5" only as default (to limit the duplicate request queueing up when DN is slow). and handle memory in queue.

Considering parallel HB and SCM deleteBlock processing,  they are synchronized using lock, 
- HB take lock, get commands, process HB response, release lock
- SCM deleteBlock, getCommandQueueCount() take lock, get count and then release lock

So below sequence,
Scenario 1: - No issue as queue empty and next command can be added
1. HB comes first
2. SCM delete block processing next

Scenario 2: Here, adding same command will be duplicate and there is retry
1. SCM delete block processing first
2. HB comes next

Considering this, we need not have queue at SCM and above changes not required.